sources/host-ctr/cmd/host-ctr/main.go

```diff
@@ -98,172 +150,177 @@ func _main() int {
 		}
 	}(ctx, cancel)
 
-	// Setup containerd client using provided socket.
-	client, err := containerd.New(containerdSocket, containerd.WithDefaultNamespace(namespace))
+	client, err := newContainerdClient(ctx, containerdSocket, namespace)
 	if err != nil {
-		log.G(ctx).
-			WithError(err).
-			WithField("socket", containerdSocket).
-			WithField("namespace", namespace).
-			Error("Failed to connect to containerd")
-		return 1
+		return err
 	}
 	defer client.Close()
 
 	// Parse the source ref if it looks like an ECR ref.
-	ref := source
-	isECRImage := ecrRegex.MatchString(ref)
+	isECRImage := ecrRegex.MatchString(source)
+	var img containerd.Image
 	if isECRImage {
-		ecrRef, err := ecr.ParseImageURI(ref)
+		img, err = pullECRImage(ctx, source, client)
 		if err != nil {
-			log.G(ctx).WithError(err).WithField("source", source).Error("Failed to parse ECR reference")
-			return 1
+			return err
+		}
+	} else {
+		img, err = pullImage(ctx, source, client)
+		if err != nil {
+			log.G(ctx).WithField("ref", source).Error(err)
+			return err
 		}
-		ref = ecrRef.Canonical()
-		log.G(ctx).
-			WithField("source", source).
-			WithField("ref", ref).
-			Debug("Parsed ECR reference from URI")
 	}
 
-	img, err := pullImage(ctx, ref, client)
+	// Check if the target container already exists. If it does, take over the helm to manage it.
```

Comment:
Can we separate the functional changes from the refactoring changes?

Reply:
Oh oops. I totally forgot that this came along with my cherry pick.